crates/transpiler/src/transpile_layout.rs

```diff
@@ -0,0 +1,493 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2025
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use qiskit_circuit::nlayout::{NLayout, PhysicalQubit, VirtualQubit};
+
+/// The "layout" caused by transpilation
+///
+/// In general Qiskit's transpiler is unitary-preserving up to the initial layout
+/// and routing permutations. The initial layout permutation is caused by
+/// setting and applying the initial layout (the mapping from virtual circuit
+/// qubits to physical qubits on the target) and the routing permtuations are
+/// caused by swap gate insertion or permutation ellision prior to the initial
+/// layout. This struct tracks these details and provide an interface to reason
+/// about these permutations.
+pub struct TranspileLayout {
+    initial_layout: NLayout,
+    routing_permutation: Option<Vec<PhysicalQubit>>,
+    input_qubit_count: u32,
+    output_qubit_count: u32,
+}
```

Comment:
This could do with more documentation, especially `routing_permutation` - this needs to clearly state how the permutation of `routing_permutation` is supposed to be interpreted.

Under what circumstances can `output_qubit_count` be different to the length of `initial_layout`?